crypto/params.c

```diff
@@ -0,0 +1,401 @@
+    switch (p->data_type) {
+    case OSSL_PARAM_REAL:
+        switch (p->buffer_size) {
+        CASE(double);
+        }
+        break;
+    case OSSL_PARAM_INTEGER:
+        switch (p->buffer_size) {
+        CASE(int32_t);
+        CASE(int64_t);
+        }
+        break;
+    case OSSL_PARAM_UNSIGNED_INTEGER:
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+        break;
+    }
+    return 0;    
+#undef CASE
+}
+
+/*******************************************************************
+* Track the allocated pointers using a sparse array.               *
+* This needs to be replaced with something better & more portable. *
+*******************************************************************/
+#include "internal/sparse_array.h"
+DEFINE_SPARSE_ARRAY_OF(void);
+static SPARSE_ARRAY_OF(void) *allocated_params = NULL;
+
+static void ossl_params_cleanup(void)
+{
+    ossl_sa_void_free(allocated_params);
+    allocated_params = NULL;
+}
+
+static CRYPTO_ONCE params_init_flag = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_params_init)
+{
+    allocated_params = ossl_sa_void_new();
+    return allocated_params != NULL && OPENSSL_atexit(&ossl_params_cleanup);
+}
+
+static void ossl_param_register_alloc(void *string)
+{
+    if (!RUN_ONCE(&params_init_flag, do_params_init))
+        return;
+
+    /* Magic to go here */
+    ossl_sa_void_set(allocated_params, (size_t)string, string);
+}
+
+void OSSL_PARAM_string_free(void *string)
+{
+    if (!RUN_ONCE(&params_init_flag, do_params_init))
+        return;
+
+    /* Magic to go here */
+    if (ossl_sa_void_get(allocated_params, (size_t)string) == NULL)
+        return;
+    ossl_sa_void_set(allocated_params, (size_t)string, NULL);
+    free(string);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, double *buf,
+                                            size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING, buf, bsize,
+                                rsize);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string_ptr(const char *key, double *buf,
+                                                size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING_PTR, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_utf8_string(const OSSL_PARAM *p, char **val, size_t max_len)
+{
+    char *q;
+
+    if (val == NULL || p == NULL)
+        return 0;
+    if (p->data_type != OSSL_PARAM_UTF8_STRING &&
+        p->data_type != OSSL_PARAM_UTF8_STRING_PTR)
+        return 0;
+
+    q = *val;
+    if (q == NULL) {
+        if (p->data_type == OSSL_PARAM_UTF8_STRING_PTR) {
+            *val = *(char **)p->buffer;
+            return 1;
+        }
+        if ((q = OPENSSL_malloc(p->buffer_size)) == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, p->buffer_size);
+        ossl_param_register_alloc(q);
```

Comment:
I think this is a bit over the top. Presumably, the calling code knows or can easily check the data type, and deduce that if it passed a reference to NULL and the passed pointer is non-NULL after the call to this function, it needs to free what that pointer points to. This should not be hard to document.

Reply:
Not quite.  A UTF8_STRING_PTR will replace a NULL with a pointer that can't be freed.
Either an extra return argument is required to indicate free the pointer or the params code has to keep track of which ones need freeing.

Reply:
I also need to do a copy for the _PTR case which isn't being done currently.

Reply:
But surely, the code that constructed the OSSL_PARAM to begin with *knows* if that OSSL_PARAM has the data type OSSL_PARAM_UTF8_STRING or OSSL_PARAM_UTF8_STRING_PTR? It's not like those data types find themselves there randomly by magic. That should be very easy to code. Do you want an example?

Reply:
Okay, I'm fine delegating the free management to the caller :)
No need for an example.

Reply:
Remember to document the requirements

Reply:
Lots of documentation to write still...

Reply:
While I'm not religiously opposed to leaving free management to the caller, it is something that should be pretty rare, as it's easy for the caller to mess it up.  As much as possible, we should stick to simple API contracts, where (e.g.) "you should always free() this parameter after calling this function" lets the calling code be simpler.

Reply:
You cannot allocate something in a DLL and free it in main because that won't work on Windows for example.

Reply:
The allocations are not across provider boundaries.  The provider calls get to extract information that has already passed the boundary.  The provider is required to call free.

Reply:
@kaduk is "you must free if you didn't provide the space" sufficiently simple?

Reply:
It is tolerably simple, yes.  May or may not be the best solution in any given case, though.